src/signature.js

```diff
@@ -0,0 +1,128 @@
+import crypto from "crypto";
+import fetch from "node-fetch";
+
+import { account, domain } from "./util.js";
+import { getPrivateKey } from "./activity-pub-db.js";
+
+/**
+ * Returns base-64 encoded SHA-256 digest of provided data
+ *
+ * @param {string} data - UTF-8 string to be hashed
+ *
+ * @returns {string}
+ */
+function getDigest(data) {
+  return crypto.createHash("sha256").update(data).digest("base64");
+}
+
+/**
+ * Returns base-64 encoded string signed with user's RSA private key
+ *
+ * @param {string} privkey - Postmarks user's private key
+ * @param {string} data - UTF-8 string to sign
+ *
+ * @returns {string}
+ */
+function getSignature(privkey, data) {
+  const signer = crypto.createSign("sha256");
+  signer.update(data);
+  signer.end();
+  return signer.sign(privkey).toString("base64");
+}
+
+/**
+ * Returns object of params to be used for HTTP signature
+ *
+ * @param {BodyInit | null} [body] - Request body for signature digest (usually a JSON string, optional)
+ * @param {string} method - Request HTTP method name
+ * @param {string} url -
+ *
+ * @returns {Object}
+ */
+function getSignatureParams(body, method, url) {
+  const urlObj = new URL(url);
+  const path = `${urlObj.pathname}${urlObj.search}`;
+  const requestTarget = `${method.toLowerCase()} ${path}`;
+  const hostParam = urlObj.hostname;
+
+  const date = new Date();
+  const dateParam = date.toUTCString();
+
+  const params = {
+    "(request-target)": requestTarget,
+    host: hostParam,
+    date: dateParam,
+  };
+
+  // add digest param if request body is present
+  if (body) {
+    const digest = getDigest(body);
+    const digestParam = `SHA-256=${digest}`;
+    params["digest"] = digestParam;
+  }
+
+  return params;
+}
+
+/**
+ * Returns the full "Signature" header to be included in the signed request
+ *
+ * @param {string} signature - Base-64 encoded request signature
+ * @param {string[]} signatureKeys - Array of param names used when generating the signature
+ *
+ * @returns {string}
+ */
+function getSignatureHeader(signature, signatureKeys) {
+  return [
+    `keyId="https://${domain}/u/${account}"`,
+    `algorithm="rsa-sha256"`,
+    `headers="${signatureKeys.join(" ")}"`,
+    `signature="${signature}"`,
+  ].join(",");
+}
+
+/**
+ * Signs a fetch request with the account's RSA private key
+ *
+ * @param {URL | RequestInfo} url - URL (passed to fetch)
+ * @param {RequestInit} [init={}] - Optional fetch init object
+ *
+ * @returns {Promise<Response>}
+ */
+export async function signedFetch(url, init = {}) {
+  const privkey = await getPrivateKey(`${account}@${domain}`);
+  if (!privkey) {
+    throw new Error(`No private key found for ${account}.`);
+  }
+
+  const { headers = {}, body = null, ...rest } = init;
+  const { method = body ? "POST" : "GET" } = init; // guess method if not provided
+  const isJSON = body && typeof body === "string"; // assume string body is JSON
```

Comment:
I wanted `signedFetch()` to match the function signature of regular `fetch()`, with the ability to pass additional options through the `init` param.

At the same time, it seemed like a good idea to assume some defaults for how it will be used in this project, so we don't have to (for example) manually specify the "Accept" header when we will almost always need it to be "application/json."

So these lines make some assumptions:
- If you don't specify `init.method`, it will assume POST if `init.body` is set, or GET if it is not.
- If `init.body` is a string, assume we are sending JSON and set the "Content-Type" header (line 110 below).

But now I'm second-guessing and think maybe it would be better to match fetch (method = "GET" unless specified, headers and everything else must be manually specified) -- but export two more functions `signedGetJSON()` and `signedPostJSON()` that just pass the correct settings to `signedFetch()`.

Reply:
I like adding those two utility functions! I agree that the naming of the base function would make me assume I needed to do everything just like I would with normal `fetch`.

It's maybe possible that this detail makes it worth consider using something with more defaults like Axios? I'm not an experienced enough JS developer to know all the tradeoffs there; it seems more popular with JS folks I've worked with, but to me I like the idea of being able to use built-in fetch one day as part of the general ethos of limiting dependencies on the project.

Reply:
I don't have experience with Axios, but I suspect at this point -- with Node 20 adding experimental support for fetch natively -- just using vanilla fetch is the way to go.